### PR TITLE
Add option to seed random number engines

### PR DESCRIPTION
**Pull request by Kai-Striega**
This PR adds an additional option `execution.seed` to explicitly seed the random number generators used by `galaxy_mergers` and `dark_matter_halos`.  

Having a seed allows single-threaded runs of SHArk to be reproducible. If not explicitly set, a random seed is generated by ``std::device_random``. This number will be non-deterministic and runs will no longer be reproducible.

The PR includes automated tests, running on both AppVoyer and Travis-CI.

### CODE REVIEW

**Comment by rtobar** [comment]
Kai, can you please merge once more from the `master` branch? I updated some small details, including a build failure in AppVeyor that started happening now.

Also, I also thought it would be good to add note in `doc/changelog.rst` about the work that is being implemented in this pull request so it appears in the online documentation.

**Comment by rtobar** [comment]
Re-opening; it seems this got automatically closed when you deleted and then restored your branch

**Comment by Kai-Striega** [comment]
@rtobar would you please take a look at this? The tests are passing for AppVoyer, XCode and TravisCI (g++6, python) but failing on TravisCI (g++7, python3).

**Comment by rtobar** [comment]
Hi Kai,

Thanks for the new batch of patches, it looks really good.

The failure in Travis with python 3 seems fairly consistent, as it failed two times with the same diagnose. This seems to be related to how the `keys` method of h5py `Group` objects behave: in python 2 it returns a list, but in python 3 it returns view-like objects (http://docs.h5py.org/en/stable/high/group.html#dict-interface-and-links). These are probably not directly comparable depending on the h5py version, hence the failure in Travis (h5py 2.2.0) but the success on AppVeyor (h5py 2.9.0). This is a guess, but I think would explain the error.

I'd suggest you `set` what's coming out of `keys()` and compare those sets to ensure compatibility across python and h5py versions. Could you give that a try? It could be something like:

```python
k1, k2 = map(set, [galaxy1.keys(), galaxy2.keys()])
if k1 != k2:
    ...
```

**Comment by rtobar** [comment]
Ohh, and one other thing I just realized (sorry!): when writing the `galaxies.hdf5` file, could you write `run_info/seed` as a dataset rather than an attribute, and add a comment to it? We currently write all the other pieces of information as datasets (except `run_info/model`, which I will turn *also* into a dataset), since they allow to add additional comments for built-in documentation.

After you do this, could you re-generate the `doc/hdf5_properties/galaxies.rst` file too? This should be doable by running `scripts/properties_as_list.sh` against a `galaxies.hdf5` file that contains the new information. In principle that should produce a one-line diff, but if for some reason the diff becomes too wild then I'd rather update the doc by hand. And while you are at it, please also add a note to `doc/changelog.rst` about the new feature you are adding. Thanks!

File: scripts/compare_galaxies.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def assert_galaxies_not_equal(galaxy1, galaxy2):
+    """Raise an AssertionError if two galaxies are equal."""
+    try:
+        assert_galaxies_equal(galaxy1, galaxy2)
+    except AssertionError as e:
+        pass
+    else:
+        raise ValueError('Galaxies expected to be unequal, but are equal.')
```
Comment (rtobar) [inline scripts/compare_galaxies.py]: Maybe use `AssertionError` here too for symmetry

File: scripts/compare_galaxies.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import argparse
+
+import h5py
+from numpy import array_equal
+
+
+def read_args():
+    """Return an argparse.Namespace object with the CLI arguments"""
+    arg_parser = argparse.ArgumentParser(
+        "Test whether two SCHARK models are equivalent."
```
Comment (rtobar) [inline scripts/compare_galaxies.py]: `shark` (mind the spelling)

File: scripts/compare_galaxies.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Test two SHARK galaxies for equality
```
Comment (rtobar) [inline scripts/compare_galaxies.py]: Preferably `shark` in lowercase, as it isn't an acronym actually

**Review by rtobar — COMMENTED** [review]
And other very minor, aesthetic comments; otherwise all checks are passing now

**Comment by rtobar** [comment]
Thanks again Kai for this pull request, I've just merged it into the `master` branch so now it's officially out there. I took the liberty of updating the documentation bit that was missing, which was a very minor thing really.

**Comment by Kai-Striega** [comment]
@rtobar thanks for taking the time to mentor/review my work.

**Pull request merged**